nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+#[derive(Debug)]
+pub struct MMError {
+    error_kind: MMErrorKind,
+    message: String,
+}
+
+/// Errors produced by functions that expect well-formed matrix market format data.
+/// > _NOTE1:_ Since the matrix market design didn't mention if duplicate entries are allowed or not, so, it's allowed here.
+///
+/// > _NOTE2:_ Dense matrices are not supported here. For example, `%%matrixmarket matrix array real general` will give `MMErrorKind::ParsingError`.
+#[non_exhaustive]
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum MMErrorKind {
+    /// Indicates that some word is not known to MM format
+    ///
+    /// Examples
+    /// --------
+    /// ```rust
+    /// # use nalgebra_sparse::io::load_coo_from_mm_str;
+    /// # use nalgebra_sparse::io::MMErrorKind;
+    /// let str = r#"
+    /// %%MatrixMarket whatever whatever whatever whatever
+    /// 1 1 1
+    /// 1 1 5
+    /// "#;
+    /// let matrix_error = load_coo_from_mm_str::<f64>(str);
+    /// assert_eq!(matrix_error.is_err(), true);
+    /// assert_eq!(*matrix_error.unwrap_err().kind(),MMErrorKind::ParsingError);
+    /// ```
+    ParsingError,
+
+    /// Indicates that header is not valid
+    ///
+    /// Examples
+    /// --------
+    /// ```rust
+    /// # use nalgebra_sparse::io::load_coo_from_mm_str;
+    /// # use nalgebra_sparse::io::MMErrorKind;
+    /// let str = r#"
+    /// %%MatrixMarket matrix coordinate real hermitian
+    /// % a real matrix can't be hermitian
+    /// 1 1 1
+    /// 1 1 5
+    /// "#;
+    /// let matrix_error = load_coo_from_mm_str::<f64>(str);
+    /// assert_eq!(matrix_error.is_err(), true);
+    /// assert_eq!(*matrix_error.unwrap_err().kind(),MMErrorKind::InvalidHeader);
+    /// ```
+    InvalidHeader,
+
+    /// Indicates that the data entries in .mtx file are more or less than entries specified in .mtx file
+    ///
+    /// Examples
+    /// --------
+    /// ```rust
+    /// # use nalgebra_sparse::io::load_coo_from_mm_str;
+    /// # use nalgebra_sparse::io::MMErrorKind;
+    /// let str = r#"
+    /// %%matrixmarket matrix coordinate real general
+    /// % it has one more data entry than specified.
+    /// 3 3 1
+    /// 2 2 2
+    /// 2 3 2
+    /// "#;
+    /// let matrix_error = load_coo_from_mm_str::<f64>(str);
+    /// assert_eq!(matrix_error.is_err(), true);
+    /// assert_eq!(*matrix_error.unwrap_err().kind(),MMErrorKind::EntryNumUnmatched);
+    /// ```
+    EntryNumUnmatched,
+
+    /// Indicates that the type T is not matched with the function it called.
+    ///
+    /// Examples
+    /// --------
+    /// ```rust
+    /// # use nalgebra_sparse::io::load_coo_from_mm_str;
+    /// # use nalgebra_sparse::io::MMErrorKind;
+    /// let str = r#"
+    /// %%matrixmarket matrix coordinate integer general
+    /// % it should be called by load_coo_from_mm_str::<i32>(str), or any other integer type;
+    /// 3 3 1
+    /// 2 2 2
+    /// 2 3 2
+    /// "#;
+    /// let matrix_error = load_coo_from_mm_str::<f64>(str);
```

Comment:
It seems to me that it wouldn't be unreasonable to load an integer matrix into a floating-point representation, provided we can fit all the integers in the floating-point type. Thoughts?

Reply:
So here should be some kind of softer restrictions? 
for example, an integer matrix can also be read as `f64` and `Complex<f64>`, and real matrix can be read by `Complex<f64>` as well.
And only given this error, when trying to read a "big" matrix type with a "small" data type? For example, reading an `f64` matrix with `load_coo_from_mm_str::<i32>`

Reply:
Yeah, I think that's a good idea. Hmm, maybe we should actually allow even reading floating-point matrices as `i32` *if* all numbers actually fit. For example, a matrix might have all integers - then perhaps we should actually allow this?

Reply:
I don't think there is a easy way to allow reading floating-point matrices as `i32`. 

First, there is no simple way to detect all float numbers are all actually int numbers, unless read the whole data once, then read it again to construct the matrix. This can lead to a huge waste. 
Second another way to do this, using `as` in rust, though all float numbers can be converted into int numbers. But it will truncate the floating part, for example 3.12 -> 3, and I don't think this a good idea either.
And it will be more complicated, if even allow reading complex matrix as `i32` (if they are has 0 as imaginary part, and real part is also a int number), for the same idea.

Reply:
Ah, I thought e.g. `i32: TryFrom<f64>` existed with exactly the behavior I suggested. Apparently `TryFrom<float> for int` generally is not available, which makes the whole thing pretty difficult.

Unfortunately it seems there also is no implementation for `TryFrom<i32> for f64` or similar - I somehow thought this existed. All right, let's stick with the current behavior. `as` casts are too error-prone, I think.